Add stricter ruff rules and helper scripts

antirotor opened the pull request:
## Changelog Description
This is adding stricter Ruff linting rules and some helper infrastructure. It is also refactoring existing code to match the new rules. Helper scripts are written in PowerShell and Bash can be used to automate tasks run on the addon repo. This is also updating GitHub ruff action, defining specific ruff version do be in sync with the GH action and switching from poetry to uv.

This is implementing https://github.com/ynput/ayon-core/issues/927

antirotor committed:
:sparkles: add pyproject and lint everything

antirotor committed:
:sparkles: add uv tool scripts

iLLiCiTiT commented on package.py:
```diff
@@ -1,3 +1,6 @@
+"""Add-on package configuration file."""
```
Can we not change this file please?

iLLiCiTiT commented on create_package.py:
```diff
@@ -166,13 +171,12 @@ def find_files_in_subdir(
     if ignore_dir_patterns is None:
         ignore_dir_patterns = IGNORE_DIR_PATTERNS
-    output: List[Tuple[str, str]] = []
+    output: list[Tuple[str, str]] = []
```
```suggestion
    output: list[FileMapping] = []
```

iLLiCiTiT commented on create_package.py:
```diff
@@ -20,19 +20,21 @@
 Package contains server side files directly,
```
How much of the changes were "must have" and "I want them"? Because to be honest I don't like most of them.

antirotor replied:
Not sure how to answer this. This file produces a lot of linter errors for me. So I can either put it to excluded files - that seems to me wrong - or I can fix the linter issues which is what I did. So I'll cowardly pass the decision to @philippe-ynput 🤠

iLLiCiTiT replied:
Why suddenly all the functions do expect `addon_client_dir`?

philippe-ynput replied:
I don't mind these changes but I do mind the use of \ for line continuation.
PEP-8:
> The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.
> 
> Backslashes may still be appropriate at times. For example, long, multiple with-statements could not use implicit continuation before Python 3.10, so backslashes were acceptable for that case:

iLLiCiTiT commented on create_package.py:
```diff
@@ -396,9 +435,16 @@ def create_addon_package(
 
 def main(
     output_dir: Optional[str] = None,
+    *,
```
?

antirotor replied:
https://docs.astral.sh/ruff/rules/boolean-type-hint-positional-argument/

https://docs.astral.sh/ruff/rules/boolean-default-value-positional-argument/

iLLiCiTiT commented on create_package.py:
```diff
@@ -407,27 +453,32 @@ def main(
     has_client_code = bool(ADDON_CLIENT_DIR)
     if has_client_code:
-        client_dir: str = os.path.join(CLIENT_ROOT, ADDON_CLIENT_DIR)
+        if ADDON_CLIENT_DIR:
```
?
It is checked on line 454, which is used on line 455, and here you have the same condition... Also having `client_dir` set to `CLIENT_ROOT` doesn't make any sense.

antirotor replied:
Again, same issue with **Mypy** check. It should be probably refactored to pass `ADDON_CLIENT_DIR` as an argument.

iLLiCiTiT commented on create_package.py:
```diff
@@ -407,27 +453,32 @@ def main(
     if only_client:
         if not has_client_code:
-            raise RuntimeError("Client code is not available. Skipping")
+            msg = "Client code is not available. Skipping"
```
Why?

antirotor replied:
https://docs.astral.sh/ruff/rules/raw-string-in-exception/

iLLiCiTiT commented on create_package.py:
```diff
@@ -407,27 +453,32 @@ def main(
+            client_dir = CLIENT_ROOT
         if not os.path.exists(client_dir):
-            raise RuntimeError(
+            msg = (
```
Why?

antirotor replied:
https://docs.astral.sh/ruff/rules/raw-string-in-exception/

iLLiCiTiT commented on create_package.py:
```diff
@@ -407,27 +453,32 @@ def main(
         return
 
-    log.info(f"Preparing package for {ADDON_NAME}-{ADDON_VERSION}")
+    log.info("Preparing package for %s, %s", ADDON_NAME, ADDON_VERSION)
```
Why? This is a script, not a module that is imported and used anywhere, can we please keep f-strings for readability?

antirotor replied:
If you can come with reasonable exception for the linter rules where it can be used and where not. In most use cases logging is happening in imported code. In this particular is not. Might be solved with `noqa`

iLLiCiTiT commented on ruff.toml:
```diff
@@ -33,11 +33,33 @@ line-length = 79
+    "S404",    # subprocess module is possibly insecure
+    "PLC0415", # import must be on top of the file
+    "CPY001",  # missing copyright header
+    "UP045"
+
```
```suggestion
```

antirotor replied:
This is connected to the ruff version. I think the current (added in this PR) needs this.

iLLiCiTiT commented on server/__init__.py:
```diff
@@ -1,13 +1,18 @@
+"""Server package."""
+from __future__ import annotations
```
I think annotations in addon might be crutial.

iLLiCiTiT commented on server/settings.py:
```diff
@@ -1,7 +1,12 @@
+"""Settings for the addon."""
+from __future__ import annotations
```
Settings models are based on type hints, you can't remove them.

antirotor replied:
They are not removed by this. Settings are Pydantic models and Pydantic is handling it as it should.

iLLiCiTiT commented on tools/manage.ps1:
```diff
@@ -0,0 +1,279 @@
+<#
```
Few notes:
- This does not belong to this PR.
- I don't think it belongs to `tools` subfolder, but should be in root, if we add it.
- The template is not used only by Ynput team, it is for community addons too.

antirotor replied:
I think these "tools" are just helpers. You can create venv just by running `uv` but this wraps another functionality around it and is convenient place to add other functionality if needed. It doesn't have to be maintained that much so I see no harm to have it there. It IMO doesn't belong to root. Also I think it doesn't really matter if this is used just by YNPUT or community, it can help both. @philippe-ynput what do you think?

philippe-ynput replied:
I agree with @iLLiCiTiT about the location but you can keep it in the PR.

iLLiCiTiT commented on create_package.py:
```diff
@@ -146,9 +151,9 @@ def _value_match_regexes(value: str, regexes: Iterable[Pattern]) -> bool:
-) -> List[Tuple[str, str]]:
+    ignore_file_patterns: Optional[list[Pattern]] = None,
+    ignore_dir_patterns: Optional[list[Pattern]] = None
+) -> list[Tuple[str, str]]:
```
```suggestion
) -> list[FileMapping]:
```

iLLiCiTiT replied:
Was resolved without changing it.

iLLiCiTiT commented on create_package.py:
```diff
@@ -207,27 +211,33 @@ def update_client_version(logger):
     yarn_executable = _get_yarn_executable()
     if yarn_executable is None:
-        raise RuntimeError("Yarn executable was not found.")
+        msg = "Yarn executable was not found."
```
Why?

antirotor replied:
https://docs.astral.sh/ruff/rules/raw-string-in-exception/

iLLiCiTiT commented on create_package.py:
```diff
@@ -248,21 +258,33 @@ def get_client_files_mapping() -> List[Tuple[str, str]]:
-        (path, os.path.join(ADDON_CLIENT_DIR, sub_path))
-        for path, sub_path in find_files_in_subdir(client_code_dir)
-    ]
+    if ADDON_CLIENT_DIR:
```
This is called only if `ADDON_CLIENT_DIR` is set. You can't get client content from `CLIENT_ROOT`. Please revert.

antirotor replied:
This is because of **Mypy** and the line below - `ADDON_CLIENT_DIR` can be `None` from the **Mypy** perspective and you can't pass `None` to `os.path.join()`. I think this check is still valid. But if `ADDON_CLIENT_DIR` can't be `None` it would be better to pass it as an argument to the function and not depending on the script-scope

iLLiCiTiT commented on create_package.py:
```diff
@@ -248,21 +258,33 @@ def get_client_files_mapping() -> List[Tuple[str, str]]:
+    """
     log.info("Preparing client code zip")
-    files_mapping: List[Tuple[str, str]] = get_client_files_mapping()
+    files_mapping: list[Tuple[str, str]] = get_client_files_mapping()
```
```suggestion
    files_mapping: list[FileMapping] = get_client_files_mapping()
```

iLLiCiTiT commented on create_package.py:
```diff
@@ -271,8 +293,15 @@ def get_client_zip_content(log) -> io.BytesIO:
+    """Get mapping of server side files to copy.
+
+    Returns:
+        list[tuple[str, str]]: List of tuples with source file and destination
```
```suggestion
        list[FileMapping]: List of tuples with source file and destination
```

antirotor committed:
:recycle: add .locks to gitignore

antirotor committed:
:recycle: fix mypy errors

antirotor committed:
:package: lock ruff version and change gh action

antirotor committed:
:fire: remove future annotations

antirotor committed:
:fire: remove public module docstring

antirotor committed:
:recycle: fix type hint

antirotor committed:
:recycle: exclude package.py from linting

because of removed public docstring in package.py that it is said shouldn't be there as that file can be parsed as text file :O

antirotor commented:
Note that **ruff-action** dropped `changed-file` options. This means ruff will run on the whole repository, not just on the files changed in the PR. @iLLiCiTiT likes it that way, I still think it should only lint what was changed: To restore this functionality with the new ruff action, this will do it:

```yaml
jobs:
  linting:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Get changed files
        id: changed-files
        uses: tj-actions/changed-files@v46
        with:
          files: |
            **.py
      - name: Lint Python files using ruff
      - uses: astral-sh/ruff-action@v3
        with:
          version-file: "pyproject.toml"
          src: ${{ steps.changed-files.outputs.all_changed_files }}
```

antirotor commented:
Can we move this somehow? Either merge it or close it?

iLLiCiTiT commented on package.py:
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
I still think we should not touch this one.

iLLiCiTiT commented on server/settings.py:
```diff
@@ -1,7 +1,10 @@
 from ayon_server.settings import BaseSettingsModel
 
-DEFAULT_VALUES = {}
+DEFAULT_VALUES: Dict[str, Any] = {}
```
```suggestion
DEFAULT_VALUES: dict[str, Any] = {}
```

iLLiCiTiT commented on server/settings.py:
```diff
@@ -1,7 +1,10 @@
+"""Settings for the addon."""
+from typing import Any, Dict
```
```suggestion
from typing import Any
```

antirotor committed:
:fire: remove type hinting

antirotor committed:
:recycle: change type hints

use newer python types in settings

philippe-ynput commented on create_package.py:
```diff
@@ -50,37 +52,37 @@
 
-VERSION_PY_CONTENT = f'''# -*- coding: utf-8 -*-
-"""Package declaring AYON addon '{ADDON_NAME}' version."""
+VERSION_PY_CONTENT = \
```
not a good backslash usage. see PEP8

philippe-ynput commented on ruff.toml:
```diff
@@ -26,18 +26,41 @@ exclude = [
+    "S404",    # subprocess module is possibly insecure
+    "PLC0415", # import must be on top of the file
+    "CPY001",  # missing copyright header
+    "UP045"
```
```suggestion
    "ANN101",  # missing-type-function-argument: for older ruff version
    "ANN204",  # missing-return-type-special-method
    "COM812",  # missing-trailing-comma
    "S603",    # subprocess-without-shell-equals-true
    "ERA001",  # commented-out-code
    "TRY003",  # raise-vanilla-args
    "UP006",   # non-pep585-annotation: older python version (type vs. Type)
    "UP007",   # non-pep604-annotation-union: ..^
    "UP035",   # deprecated-import: ..
    "UP045",   # non-pep604-annotation-optional: Use `X | None` for type annotations
    "ARG002",  # unused-method-argument
    "INP001",  # implicit-namespace-package: add `__init__.py` to namespaced package
    "FIX002",  # line-contains-todo: FIX all TODOs
    "TD003",   # missing-todo-link: missing issue link
    "S404",    # suspicious-subprocess-import: subprocess module can be insecure
    "PLC0415", # import-outside-top-level: import must be on top of the file
    "CPY001",  # missing-copyright-notice: missing copyright header
```

philippe-ynput reviewed (changes requested):
@antirotor please address all comments and re-request review. We are close ! ;-)

antirotor committed:
Update create_package.py

Co-authored-by: Jakub Trllo <43494761+iLLiCiTiT@users.noreply.github.com>

antirotor committed:
:dog: replace backslash with parenthesis

antirotor committed:
sorted rules with better descriptions

Co-authored-by: Philippe Leprince <philippe@ynput.io>
